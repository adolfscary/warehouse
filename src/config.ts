export const config = {
  position: {
    url:
      process.env.POSITION_API_URL ??
      "https://dev.aux.boxpi.com/case-study/products/${productId}/positions",
    apiKey:
      process.env.POSITION_API_KEY ??
      "MVGBMS0VQI555bTery9qJ91BfUpi53N24SkKMf9Z",
  },
  server: {
    port: process.env.PORT ?? 3000,
  },
};
