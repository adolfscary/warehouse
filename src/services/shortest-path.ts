type Vertices = Record<string, { visited: boolean; productId: string }>;
type Edges = Record<string, { toEdgeId: string; distance: number }[]>;
type VisitedPath = { productId: string; positionId: string };
type TSPResult = { shortestDistance: number; shortestPath: VisitedPath[] };
type Coordinate = { x: number; y: number; z: number };

const STARTING_POSITION_ID = "#starting-pos";

const calulcateDistance = (v1: Coordinate, v2: Coordinate) => {
  const a = Math.abs(v1.x - v2.x);
  const b = Math.abs(v1.y - v2.y);
  const c = Math.abs(v1.z - v2.z);
  const d = Math.sqrt(Math.pow(a, 2) + Math.pow(b, 2) + Math.pow(c, 2));
  return d;
};

const cloneVertices = (verticies: Vertices) => {
  return JSON.parse(JSON.stringify(verticies)) as Vertices;
};

const travelingSalesmanProblem = (input: {
  edges: Edges;
  verticies: Vertices;
  currentVerticeId: string;
  visitedPath: VisitedPath[];
  countVisitedVertices: number;
  countVertices: number;
  distance: number;
  result: TSPResult;
}) => {
  const {
    edges,
    verticies,
    currentVerticeId,
    visitedPath,
    countVisitedVertices,
    countVertices,
    distance,
    result,
  } = input;
  if (countVisitedVertices === countVertices) {
    if (distance < result.shortestDistance) {
      result.shortestDistance = distance;
      result.shortestPath = visitedPath;
      return;
    }
  }
  for (const edge of edges[currentVerticeId]) {
    if (verticies[edge.toEdgeId].visited) {
      continue;
    }
    const verticesCopy = cloneVertices(verticies);
    verticesCopy[edge.toEdgeId].visited = true;
    for (const vertice of Object.values(verticesCopy)) {
      if (vertice.productId === verticesCopy[edge.toEdgeId].productId) {
        vertice.visited = true;
      }
    }
    travelingSalesmanProblem({
      ...input,
      verticies: verticesCopy,
      countVisitedVertices: countVisitedVertices + 1,
      visitedPath: [
        ...visitedPath,
        {
          productId: verticies[edge.toEdgeId].productId,
          positionId: edge.toEdgeId,
        },
      ],
      currentVerticeId: edge.toEdgeId,
      distance: distance + edge.distance,
    });
  }
};

export const findShortestPath = ({
  startingPosition,
  productsIds,
  positions,
}: {
  startingPosition: { x: number; y: number; z: number };
  productsIds: string[];
  positions: {
    positionId: string;
    productId: string;
    x: number;
    y: number;
    z: number;
  }[];
}) => {
  const verticies: Vertices = {};
  const edges: Edges = {};

  verticies[STARTING_POSITION_ID] = { visited: true, productId: "" };
  edges[STARTING_POSITION_ID] = positions.map((p) => ({
    distance: calulcateDistance(
      { x: startingPosition.x, y: startingPosition.y, z: startingPosition.z },
      { x: p.x, y: p.y, z: p.z }
    ),
    toEdgeId: p.positionId,
  }));

  for (const position of positions) {
    verticies[position.positionId] = {
      visited: false,
      productId: position.productId,
    };
    edges[position.positionId] = positions.map((p) => ({
      distance: calulcateDistance(
        { x: position.x, y: position.y, z: position.z },
        { x: p.x, y: p.y, z: p.z }
      ),
      toEdgeId: p.positionId,
    }));
  }

  const result: TSPResult = {
    shortestDistance: Infinity,
    shortestPath: [],
  };

  travelingSalesmanProblem({
    verticies,
    edges,
    currentVerticeId: STARTING_POSITION_ID,
    countVertices: productsIds.length,
    visitedPath: [],
    countVisitedVertices: 0,
    distance: 0,
    result,
  });
  return result;
};
