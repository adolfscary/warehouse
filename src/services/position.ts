import { config } from "../config";
import axios from "axios";

type Position = {
  positionId: string;
  x: number;
  y: number;
  z: number;
  productId: string;
  quantity: number;
};

const getPosition = async (productId: string): Promise<Position[]> => {
  const url = config.position.url.replace("${productId}", productId);

  const { data } = await axios({
    url,
    method: "GET",
    headers: {
      "X-API-KEY": config.position.apiKey,
    },
  });
  return data;
};

export const getPositions = async (productIds: string[]) => {
  return (
    await Promise.all(productIds.map((productId) => getPosition(productId)))
  ).flat();
};
