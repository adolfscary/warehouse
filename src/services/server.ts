import { config } from "../config";
import { getPositions } from "./position";
import { findShortestPath } from "./shortest-path";
import express, { Request, Response } from "express";
import bodyParser from "body-parser";
import { once } from "events";

const asyncHandler = (fun: any) => (req: Request, res: Response, next: any) => {
  Promise.resolve(fun(req, res, next)).catch(next);
};

export const startServer = async () => {
  const app = express();
  const port = config.server.port;

  app.use(bodyParser.json());

  app.post(
    "/",
    asyncHandler(async (req: Request, res: Response) => {
      const productsIds = req.body.products;
      const startingPosition = req.body.startingPosition;
      const positions = await getPositions(productsIds);
      const result = findShortestPath({
        startingPosition,
        productsIds,
        positions,
      });
      res.json({
        pickingOrder: result.shortestPath,
        distance: result.shortestDistance,
      });
    })
  );

  const server = app.listen(port, () => {
    console.log(`Server running at http://localhost:${port}`);
  });

  await once(server, "listening");

  return server;
};
