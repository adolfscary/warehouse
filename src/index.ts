import "dotenv/config";

import { startServer } from "./services/server";

const handles: {
  server?: Awaited<ReturnType<typeof startServer>>;
} = {};

const start = async () => {
  handles.server = await startServer();
};

start().catch(console.error);
