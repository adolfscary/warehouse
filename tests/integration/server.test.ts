import { config } from "../../src/config";
import { startServer } from "../../src/services/server";
import { once } from "events";
import { AddressInfo } from "net";

describe("integration", () => {
  it("server", async () => {
    config.server.port = 0;
    const server = await startServer();
    const port = (server.address() as AddressInfo).port;
    const res = await fetch(`http://localhost:${port}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        products: ["product-1", "product-2"],
        startingPosition: { x: 0, y: 0, z: 0 },
      }),
    });

    expect(res.ok).toBe(true);
    const json = (await res.json()) as { pickingOrder?: { length?: number } };

    expect(json).toMatchSnapshot({
      pickingOrder: Array(json?.pickingOrder?.length).fill({
        productId: expect.any(String),
        positionId: expect.any(String),
      }),
      distance: expect.any(Number),
    });

    server.close();
    server.closeAllConnections();
    await once(server, "close");
  });
});
