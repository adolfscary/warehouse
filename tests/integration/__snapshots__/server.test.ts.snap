// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration server 1`] = `
{
  "distance": Any<Number>,
  "pickingOrder": [
    {
      "positionId": Any<String>,
      "productId": Any<String>,
    },
    {
      "positionId": Any<String>,
      "productId": Any<String>,
    },
  ],
}
`;
